### PR TITLE
Query: Perf: Adds explicit compiled queries.

### PR DESCRIPTION
**Pull request by anpete**
- Adds EF.CompileQuery and EF.CompileAsyncQuery which allow for the creation of user managed
compiled query delegates. Such queries are more efficient to execute because they bypass the
CompiledQueryCache (avoiding query hashing and parameterization). Typical gains are around 3-5%

### CODE REVIEW

File: src/Microsoft.EntityFrameworkCore.Specification.Tests/CompiledQueryTestBase.cs [hunk 1]
```diff
@@ -0,0 +1,359 @@
...
+
+            using (var context = CreateContext())
+            {
+                Assert.Equal("ANATR", query(context, null, 1, "ANATR").First().CustomerID);
+            }
+        }
+
+        [ConditionalFact]
+        public virtual void Query_with_array_parameter()
+        {
```
Comment (maumar) [inline src/Microsoft.EntityFrameworkCore.Specification.Tests/CompiledQueryTestBase.cs]: another interesting query could be:

context.Customers.OrderBy(c => c.Id).Select(c => c.Name).FirstOrDefault() + context.Orders.OrderBy(o => o.Id).Select(o.CustomerId).FirstOrDefault()

i.e. the root of the query is context, not dbset, curious if that gets translated into one query or will string concat be client-evaled

File: src/Microsoft.EntityFrameworkCore/EF.CompileAsyncQuery.cs [hunk 2]
```diff
@@ -0,0 +1,304 @@
...
+        /// </summary>
+        /// <typeparam name="TContext">The target DbContext type.</typeparam>
+        /// <typeparam name="TParam1">The type of the first query parameter.</typeparam>
+        /// <typeparam name="TParam2">The type of the second query parameter.</typeparam>
+        /// <typeparam name="TParam3">The type of the third query parameter.</typeparam>
+        /// <typeparam name="TParam4">The type of the fourth query parameter.</typeparam>
+        /// <typeparam name="TParam5">The type of the fifth query parameter.</typeparam>
+        /// <typeparam name="TResult">The query result type.</typeparam>
+        /// <param name="queryExpression">The LINQ query expression.</param>
+        /// <returns>A delegate that can be invoked to execute the compiled query.</returns>
```
Comment (maumar) [inline src/Microsoft.EntityFrameworkCore/EF.CompileAsyncQuery.cs]: we think 5 parameters is enough? in EF6 we had 10+ IIRC

Comment (anpete) [inline src/Microsoft.EntityFrameworkCore/EF.CompileAsyncQuery.cs]: We can add more. I'm also going to enable object arrays as an alternative.

Comment (divega) [inline src/Microsoft.EntityFrameworkCore/EF.CompileAsyncQuery.cs]: In the old stack we had `ObjectContext` + 15 arguments + `TResult` = 17. I think 5 is a good start :smile:

File: src/Microsoft.EntityFrameworkCore/Query/Internal/CompiledAsyncTaskQuery.cs [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+using System.Threading.Tasks;
+using JetBrains.Annotations;
+
+namespace Microsoft.EntityFrameworkCore.Query.Internal
+{
+    /// <summary>
+    ///     This API supports the Entity Framework Core infrastructure and is not intended to be used
+    ///     directly from your code. This API may change or be removed in future releases.
+    /// </summary>
+    public class CompiledAsyncTaskQuery<TContext, TResult> : CompiledQueryBase<TContext, Task<TResult>>
```
Comment (maumar) [inline src/Microsoft.EntityFrameworkCore/Query/Internal/CompiledAsyncTaskQuery.cs]: nit: Make the name more indicative of what it does? (Singleton/Scalar etc) It was not obvious for me at the first glance.

File: src/Microsoft.EntityFrameworkCore/EF.CompileAsyncQuery.cs [hunk 4]
```diff
@@ -0,0 +1,304 @@
...
+    public static partial class EF
+    {
+        /// <summary>
+        ///     Creates a compiled query delegate that when invoked will execute the specified LINQ query.
+        /// </summary>
+        /// <typeparam name="TContext">The target DbContext type.</typeparam>
+        /// <typeparam name="TResult">The query result type.</typeparam>
+        /// <param name="queryExpression">The LINQ query expression.</param>
+        /// <returns>A delegate that can be invoked to execute the compiled query.</returns>
+        public static Func<TContext, IAsyncEnumerable<TResult>> CompileAsyncQuery<TContext, TResult>(
```
Comment (divega) [inline src/Microsoft.EntityFrameworkCore/EF.CompileAsyncQuery.cs]: Open issue @anpete and I already talked about: this may mean the user has to add a direct reference to `System.Interactive.Async` which `IAsyncEnumerable<T> for any meaningful usage. We may make it easier by creating a few extension methods, e.g. `ToListAsync()` and `ForEachAsync()`. Another option could be to return `IQueryable<T>` or an EF specific query result type for which we support async enumeration.

File: src/Microsoft.EntityFrameworkCore/Query/Internal/CompiledAsyncTaskQuery.cs (hunk 3 above)
Comment (anpete) [inline src/Microsoft.EntityFrameworkCore/Query/Internal/CompiledAsyncTaskQuery.cs]: Had "Singleton" but switched to Task for clarity 😄 Happy to change it though.

File: src/Microsoft.EntityFrameworkCore/EF.CompileAsyncQuery.cs (hunk 2 above)
Comment (anpete) [inline src/Microsoft.EntityFrameworkCore/EF.CompileAsyncQuery.cs]: @maumar volunteered to add more!

File: src/Microsoft.EntityFrameworkCore.Specification.Tests/CompiledQueryTestBase.cs (hunk 1 above)
Comment (anpete) [inline src/Microsoft.EntityFrameworkCore.Specification.Tests/CompiledQueryTestBase.cs]: Looks like we need to do some work to make these work. Filed #7016

File: src/Microsoft.EntityFrameworkCore.Specification.Tests/CompiledQueryTestBase.cs [hunk 5]
```diff
@@ -0,0 +1,359 @@
...
+            }
+
+            using (var context = CreateContext())
+            {
+                Assert.Equal("ANATR", query(context, null, 1, "ANATR").First().CustomerID);
+            }
+        }
+
+        [ConditionalFact]
+        public virtual void Query_with_array_parameter()
```
Comment (anpete) [inline src/Microsoft.EntityFrameworkCore.Specification.Tests/CompiledQueryTestBase.cs]: Added

**Comment by divega** [comment]
@anpete reminder to add a test that uses `FromSql()`

**Comment by roji** [comment]
It may be worth thinking about this in relation with prepared statements (#5459).

The explicit compiled queries implemented here aren't bound to any context, so they don't exactly correspond (explicit compiled queries are about caching inside EF, while prepared statements are about caching database-side). However, what would the statement preparation API look like, and how would it interact with this API? It's safe to say that these 2 features will almost always go together (if one wants to cache EF-side resources, one would probably also want to cache database-side resources).

**Comment by anpete** [comment]
@divega @maumar Added a wrapping async enumerable as we discussed.

**Comment by anpete** [comment]
@roji AFAIK, we don't need this for SQL Server because it happens auto-magically. I'm guessing there is a good reason to make it explicit opt-in for Postgres?

**Comment by roji** [comment]
@anpete yeah, SQL Server implicitly caches query plans etc. - but this isn't the standard state of affairs across databases. PostgreSQL requires you to explicitly prepare statements, and there'a a huge performance benefit associated with doing so (am pretty sure other databases exist which are similar).. ADO.NET as an API supports this via DbCommand.Prepare(), so it makes sense for EF to support preparation in at least some way.

Note that I'm looking into providing opt-in automatic query preparation in Npgsql with an LRU approach - this would make it redundant at the EF level, but it's still not done (and wouldn't affect other databases which can benefit from preparation). In fact, if you guys give some priority to preparation in EF I'll probably delay/drop the functionality in the ADO.NET later.

Basically for a strong cross-database story you really need to address this in some way...

**Comment by tinchou** [comment]
Hey, sorry to ask here but I'm investigating this feature for a presentation and couldn't find a better place :).

I've seen in [this PR](https://github.com/aspnet/benchmarks/pull/179/files#diff-580ab548c69197b6c524292dcbc3ebe8R28) in aspnet/benchmarks that you switch from using `FirstAsync` to `CompileAsyncQuery` + `First`. Does that mean that you only use sync methods when compiling queries?

**Comment by anpete** [comment]
@tinchou Correct. The Async versions cannot be used in expression trees because they have optional args 😦

**Comment by tinchou** [comment]
Bummer! But internally we're using the same paths as the async overloads and this is true async code, right? So in the end it wouldn't matter which method you chose.

**Comment by anpete** [comment]
Yeah, still async 😌
